lib/skate/detours/detours.ex

```diff
@@ -55,6 +56,39 @@ defmodule Skate.Detours.Detours do
     |> Repo.insert()
   end
 
+  @doc """
+  Creates a detour given a user id & detour id.
+  """
+  def create_detour_for_user(user_id, attrs \\ %{}) do
+    user = User.get_by_id!(user_id)
+
+    %Detour{
+      author: user
+    }
+    |> Detour.changeset(attrs)
+    |> Repo.insert()
+  end
+
+  @doc """
+  Update or create a detour given a user id & detour id.
+  """
+  def update_or_create_detour_for_user(user_id, uuid, attrs \\ %{}) do
+    user = User.get_by_id!(user_id)
+
+    case uuid do
+      nil ->
+        create_detour_for_user(user_id, attrs)
+
+      _ ->
+        Repo.insert(
+          Detour.changeset(%Detour{author: user, id: uuid}, attrs),
+          returning: true,
+          conflict_target: [:id],
+          on_conflict: {:replace, [:state, :updated_at]}
+        )
```

Comment:
issue: is `author` respected here? It seems like a user could update a detour that doesn't belong to them by injecting whatever ID they want.

I'm looking at providing a suggestion for all of this though, so you don't need to jump on this issue immediately, just wanted to bring it to your attention, maybe we write tests for this.

Reply:
I tested with this and found that a user can update a detour that does not belong to them

```elixir
    test "update_or_create_detour_for_user/3 should not update a detour made by another user" do
      detour = detour_fixture()
      other_user = insert(:user)

      {:ok, new_detour} = Detours.update_or_create_detour_for_user(
          other_user.id,
          detour.id,
          %{state: %{a: 1}}
        )

      assert detour.id == new_detour.id
      assert detour.author_id == new_detour.author_id
      assert detour.state == new_detour.state
    end
```

Reply:
Hmm yeah that's fair! but it does bring up the issue of permissions, roles and access control, because as is, this still allows someone who doesn't have the `detours-pilot` role to update a detour, if they knew the endpoint. I don't think this matters for our threat model, but maybe we need to figure out our threat model.